File: vouch/robotics/identity.py
"""
Hardware-rooted robot identity and lifecycle credential profile (Phase 5.1).

A documented profile for a robot's verifiable identity, bound to a hardware root
of trust (a TPM or a secure element), plus a reference implementation. It is the
open, vendor-neutral alternative to closed or state-run robot-ID schemes.

A RobotIdentityCredential is an eddsa-jcs-2022 VC whose subject carries the
robot's make, model, and serial, a lifecycle history, and a `hardwareRoot` block.
The hardware root signs a binding over (robot DID, robot key), so the robot's
software identity key is provably bound to a specific piece of hardware. A
verifier checks both the credential proof (the robot key) and the hardware
attestation (the root key).

The hardware root is pluggable. `SoftwareRootOfTrust` is the reference; a real
deployment substitutes a TPM- or secure-element-backed implementation that signs
with a hardware-resident attestation key. The interface is the contract such a
backend satisfies.
"""

from __future__ import annotations

import base64
from abc import ABC, abstractmethod
from dataclasses import dataclass
from datetime import datetime, timedelta, timezone
from typing import Any, Dict, List, Optional

from cryptography.exceptions import InvalidSignature
from cryptography.hazmat.primitives.asymmetric.ed25519 import (
    Ed25519PrivateKey,
    Ed25519PublicKey,
)
from cryptography.hazmat.primitives.serialization import Encoding, PublicFormat

from vouch import multikey
from vouch.jcs import canonicalize
from ._signing import attach_proof

VC_CONTEXT_V2 = "https://www.w3.org/ns/credentials/v2"
VOUCH_CONTEXT_V1 = "https://vouch-protocol.com/contexts/v1"
ROBOT_IDENTITY_TYPE = "RobotIdentityCredential"


class RoboticsError(Exception):
    """Raised on malformed robot-identity input."""


# ---------------------------------------------------------------------------
# Hardware root of trust
# ---------------------------------------------------------------------------


class HardwareRootOfTrust(ABC):
    """
    A hardware-resident key that attests the robot's software identity key.

    A TPM backend signs with an Attestation Key (AK) derived from the
    Endorsement Key; a secure-element backend signs with a device key fused at
    manufacture. Both satisfy this interface.
    """

    @abstractmethod
    def public_key_raw(self) -> bytes:
        """The 32-byte Ed25519 public key of the hardware root."""

    @abstractmethod
    def sign(self, data: bytes) -> bytes:
        """Sign `data` with the hardware-resident key."""

    @property
    def kind(self) -> str:
        return "Software"

    def public_key_multibase(self) -> str:
        return multikey.encode_ed25519_public(self.public_key_raw())


class SoftwareRootOfTrust(HardwareRootOfTrust):
    """
    Reference root of trust backed by a local Ed25519 key. Stands in for a TPM
    or secure element in development and tests. NOT a hardware root: a real
    deployment MUST use a hardware-backed implementation.
    """

    def __init__(self, seed: Optional[bytes] = None, kind: str = "Software") -> None:
        self._sk = (
            Ed25519PrivateKey.from_private_bytes(seed) if seed else Ed25519PrivateKey.generate()
        )
        self._kind = kind

    def public_key_raw(self) -> bytes:
        return self._sk.public_key().public_bytes(Encoding.Raw, PublicFormat.Raw)

    def sign(self, data: bytes) -> bytes:
        return self._sk.sign(data)

    @property
    def kind(self) -> str:
        return self._kind


# ---------------------------------------------------------------------------
# Binding and minting
# ---------------------------------------------------------------------------


def _binding_bytes(robot_did: str, robot_key_multibase: str) -> bytes:
    """Canonical bytes the hardware root signs to bind the identity key."""
    return canonicalize({"key": robot_key_multibase, "robotDid": robot_did})


def _mb64(b: bytes) -> str:
    return "u" + base64.urlsafe_b64encode(b).rstrip(b"=").decode("ascii")


def _unmb64(s: str) -> bytes:
    if not s.startswith("u"):
        raise RoboticsError("expected multibase 'u' prefix")
    payload = s[1:]
    return base64.urlsafe_b64decode(payload + "=" * (-len(payload) % 4))


def mint_robot_identity(
    robot_signer: Any,
    root: HardwareRootOfTrust,
    *,
    make: str,
    model: str,
    serial: str,
    owner: Optional[str] = None,
    lifecycle: Optional[List[Dict[str, Any]]] = None,
    valid_seconds: Optional[int] = None,
    valid_from: Optional[datetime] = None,
) -> Dict[str, Any]:
    """
    Mint a hardware-attested RobotIdentityCredential. The robot self-issues the
    credential with its Vouch key (`robot_signer`); the hardware root signs a
    binding over the robot DID and key, embedded as `hardwareRoot.attestation`.
    """
    robot_did = robot_signer.get_did()
    robot_key_mb = robot_signer.get_public_key_multikey()
    attestation = root.sign(_binding_bytes(robot_did, robot_key_mb))

    issued = (valid_from or datetime.now(timezone.utc)).astimezone(timezone.utc)
    subject: Dict[str, Any] = {
        "id": robot_did,
        "make": make,
        "model": model,
        "serial": serial,
        "hardwareRoot": {
            "kind": root.kind,
            "publicKeyMultibase": root.public_key_multibase(),
            "attestation": _mb64(attestation),
        },
        "lifecycle": lifecycle
        or [
            {"event": "commissioned", "timestamp": _iso(issued)},
        ],
    }
    if owner is not None:
        subject["owner"] = owner

    credential: Dict[str, Any] = {
        "@context": [VC_CONTEXT_V2, VOUCH_CONTEXT_V1],
        "type": ["VerifiableCredential", ROBOT_IDENTITY_TYPE],
        "issuer": robot_did,
        "validFrom": _iso(issued),
        "credentialSubject": subject,
    }
    if valid_seconds is not None:
        credential["validUntil"] = _iso(issued + timedelta(seconds=valid_seconds))
    return attach_proof(credential, robot_signer)


def verify_robot_identity(
    credential: Dict[str, Any],
    robot_public_key: Any,
) -> "tuple[bool, Optional[Dict[str, Any]]]":
    """
    Verify a RobotIdentityCredential: the credential proof (robot key) AND the
    hardware-root attestation binding the robot key to the hardware. Returns
    (ok, credentialSubject).
    """
    from vouch import data_integrity
    from vouch.verifier import _coerce_ed25519_public_key

    type_field = credential.get("type") or []
    if isinstance(type_field, str):
        type_field = [type_field]
    if ROBOT_IDENTITY_TYPE not in type_field:
        return False, None

    resolved = (
        _coerce_ed25519_public_key(robot_public_key) if robot_public_key is not None else None
    )
    if resolved is None:
        return False, None
    try:
        if not data_integrity.verify_proof(credential, resolved):
            return False, None
    except ValueError:
        return False, None

    subject = credential.get("credentialSubject") or {}
    hw = subject.get("hardwareRoot") or {}
    hw_mb = hw.get("publicKeyMultibase")
    attestation = hw.get("attestation")
    if not hw_mb or not attestation:
        return False, None

    try:
        alg, hw_raw = multikey.decode(hw_mb)
        if alg != "Ed25519":
            return False, None
        hw_pub = Ed25519PublicKey.from_public_bytes(hw_raw)
        robot_raw = resolved.public_bytes(Encoding.Raw, PublicFormat.Raw)
        robot_key_mb = multikey.encode_ed25519_public(robot_raw)
        binding = _binding_bytes(subject.get("id", ""), robot_key_mb)
        hw_pub.verify(_unmb64(attestation), binding)
    except (InvalidSignature, RoboticsError, ValueError):
        return False, None

    return True, subject


def lifecycle_event(
    event: str,
    *,
    actor: Optional[str] = None,
    details: Optional[Dict[str, Any]] = None,
    timestamp: Optional[datetime] = None,
) -> Dict[str, Any]:
    """Build a lifecycle history entry (manufactured, commissioned, transferred, decommissioned, ...)."""
    entry: Dict[str, Any] = {
        "event": event,
        "timestamp": _iso(timestamp or datetime.now(timezone.utc)),
    }
    if actor is not None:
        entry["actor"] = actor
    if details is not None:
        entry["details"] = details
    return entry


def _iso(dt: datetime) -> str:
    return dt.astimezone(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ")


__all__ = [
    "ROBOT_IDENTITY_TYPE",
    "RoboticsError",
    "HardwareRootOfTrust",
    "SoftwareRootOfTrust",
    "mint_robot_identity",
    "verify_robot_identity",
    "lifecycle_event",
]

File: packages/sdk-ts/src/robotics/identity.ts
/**
 * Hardware-rooted robot identity (TypeScript), Phase 5.1.
 *
 * Mirrors `vouch/robotics/identity.py` with byte-identical output. A
 * RobotIdentityCredential is an eddsa-jcs-2022 VC whose subject carries the
 * robot's make, model, and serial, a lifecycle history, and a `hardwareRoot`
 * block. The hardware root signs a binding over (robot DID, robot key), so the
 * software identity key is provably bound to a specific piece of hardware.
 * Verification checks both the credential proof (robot key) and the hardware
 * attestation (root key). Cross-language interop with the Python module is
 * REQUIRED: both produce byte-identical canonical bytes and verify each other.
 */

import * as crypto from 'crypto';

import { verifyProof } from '../data-integrity';
import { canonicalize } from '../jcs';
import { decode, encodeEd25519Public } from '../multikey';
import type { Signer } from '../signer';

export const VC_CONTEXT_V2 = 'https://www.w3.org/ns/credentials/v2';
export const VOUCH_CONTEXT_V1 = 'https://vouch-protocol.com/contexts/v1';
export const ROBOT_IDENTITY_TYPE = 'RobotIdentityCredential';

export class RoboticsError extends Error {}

// Node needs DER-wrapped keys; raw Ed25519 keys get these fixed prefixes.
const PKCS8_ED25519_PREFIX = Buffer.from('302e020100300506032b657004220420', 'hex');
const SPKI_ED25519_PREFIX = Buffer.from('302a300506032b6570032100', 'hex');

function privKeyFromSeed(seed: Uint8Array): crypto.KeyObject {
  if (seed.length !== 32) throw new RoboticsError('Ed25519 seed must be 32 bytes');
  return crypto.createPrivateKey({
    key: Buffer.concat([PKCS8_ED25519_PREFIX, Buffer.from(seed)]),
    format: 'der',
    type: 'pkcs8',
  });
}

function pubKeyFromRaw(raw: Uint8Array): crypto.KeyObject {
  return crypto.createPublicKey({
    key: Buffer.concat([SPKI_ED25519_PREFIX, Buffer.from(raw)]),
    format: 'der',
    type: 'spki',
  });
}

function rawPublicOf(key: crypto.KeyObject): Uint8Array {
  const jwk = key.export({ format: 'jwk' }) as { x?: string };
  if (!jwk.x) throw new RoboticsError('cannot extract Ed25519 public key bytes');
  return new Uint8Array(Buffer.from(jwk.x, 'base64url'));
}

function mb64(b: Uint8Array): string {
  return 'u' + Buffer.from(b).toString('base64url');
}

function unmb64(s: string): Uint8Array {
  if (!s.startsWith('u')) throw new RoboticsError("expected multibase 'u' prefix");
  return new Uint8Array(Buffer.from(s.slice(1), 'base64url'));
}

/** Canonical bytes the hardware root signs to bind the identity key. */
function bindingBytes(robotDid: string, robotKeyMultibase: string): Uint8Array {
  return canonicalize({ key: robotKeyMultibase, robotDid });
}

function iso(d: Date): string {
  return d.toISOString().replace(/\.\d{3}Z$/, 'Z');
}

// ---------------------------------------------------------------------------
// Hardware root of trust
// ---------------------------------------------------------------------------

export interface HardwareRootOfTrust {
  readonly kind: string;
  publicKeyRaw(): Uint8Array;
  sign(data: Uint8Array): Uint8Array;
  publicKeyMultibase(): string;
}

/**
 * Reference root of trust backed by a local Ed25519 key. Stands in for a TPM or
 * secure element in development and tests. NOT a hardware root: a real
 * deployment MUST use a hardware-backed implementation.
 */
export class SoftwareRootOfTrust implements HardwareRootOfTrust {
  readonly kind: string;
  private sk: crypto.KeyObject;
  private pkRaw: Uint8Array;

  constructor(seed?: Uint8Array, kind = 'Software') {
    this.sk = seed ? privKeyFromSeed(seed) : crypto.generateKeyPairSync('ed25519').privateKey;
    this.pkRaw = rawPublicOf(crypto.createPublicKey(this.sk));
    this.kind = kind;
  }

  publicKeyRaw(): Uint8Array {
    return this.pkRaw;
  }

  sign(data: Uint8Array): Uint8Array {
    return new Uint8Array(crypto.sign(null, Buffer.from(data), this.sk));
  }

  publicKeyMultibase(): string {
    return encodeEd25519Public(this.pkRaw);
  }
}

// ---------------------------------------------------------------------------
// Mint and verify
// ---------------------------------------------------------------------------

export interface MintRobotIdentityOptions {
  make: string;
  model: string;
  serial: string;
  owner?: string;
  lifecycle?: Array<Record<string, unknown>>;
  validSeconds?: number;
  validFrom?: Date;
}

/**
 * Mint a hardware-attested RobotIdentityCredential. The robot self-issues with
 * its Vouch key (`robotSigner`); the hardware root signs a binding over the
 * robot DID and key, embedded as `hardwareRoot.attestation`.
 */
export async function mintRobotIdentity(
  robotSigner: Signer,
  root: HardwareRootOfTrust,
  opts: MintRobotIdentityOptions
): Promise<Record<string, unknown>> {
  const robotDid = robotSigner.getDid();
  const robotKeyMb = await robotSigner.getPublicKeyMultikey();
  const attestation = root.sign(bindingBytes(robotDid, robotKeyMb));

  const issued = opts.validFrom ?? new Date();
  const subject: Record<string, unknown> = {
    id: robotDid,
    make: opts.make,
    model: opts.model,
    serial: opts.serial,
    hardwareRoot: {
      kind: root.kind,
      publicKeyMultibase: root.publicKeyMultibase(),
      attestation: mb64(attestation),
    },
    lifecycle: opts.lifecycle ?? [{ event: 'commissioned', timestamp: iso(issued) }],
  };
  if (opts.owner !== undefined) subject.owner = opts.owner;

  const credential: Record<string, unknown> = {
    '@context': [VC_CONTEXT_V2, VOUCH_CONTEXT_V1],
    type: ['VerifiableCredential', ROBOT_IDENTITY_TYPE],
    issuer: robotDid,
    validFrom: iso(issued),
    credentialSubject: subject,
  };
  if (opts.validSeconds !== undefined) {
    credential.validUntil = iso(new Date(issued.getTime() + opts.validSeconds * 1000));
  }
  return robotSigner.attachProof(credential);
}

/**
 * Verify a RobotIdentityCredential: the credential proof (robot key) AND the
 * hardware-root attestation binding the robot key to the hardware.
 */
export function verifyRobotIdentity(
  credential: Record<string, unknown>,
  robotPublicKey: crypto.KeyObject
): { ok: boolean; subject?: Record<string, unknown> } {
  const typeField = credential.type;
  const types = Array.isArray(typeField) ? typeField : [typeField];
  if (!types.includes(ROBOT_IDENTITY_TYPE)) return { ok: false };

  try {
    if (!verifyProof(credential, robotPublicKey)) return { ok: false };
  } catch {
    return { ok: false };
  }

  const subject = (credential.credentialSubject ?? {}) as Record<string, unknown>;
  const hw = (subject.hardwareRoot ?? {}) as Record<string, unknown>;
  const hwMb = hw.publicKeyMultibase as string | undefined;
  const attestation = hw.attestation as string | undefined;
  if (!hwMb || !attestation) return { ok: false };

  try {
    const { algorithm, rawKey } = decode(hwMb);
    if (algorithm !== 'Ed25519') return { ok: false };
    const hwPub = pubKeyFromRaw(rawKey);
    const robotRaw = rawPublicOf(robotPublicKey);
    const robotKeyMb = encodeEd25519Public(robotRaw);
    const binding = bindingBytes((subject.id as string) ?? '', robotKeyMb);
    const ok = crypto.verify(null, Buffer.from(binding), hwPub, Buffer.from(unmb64(attestation)));
    if (!ok) return { ok: false };
  } catch {
    return { ok: false };
  }
  return { ok: true, subject };
}

/** Build a lifecycle history entry. */
export function lifecycleEvent(
  event: string,
  opts: { actor?: string; details?: Record<string, unknown>; timestamp?: Date } = {}
): Record<string, unknown> {
  const entry: Record<string, unknown> = {
    event,
    timestamp: iso(opts.timestamp ?? new Date()),
  };
  if (opts.actor !== undefined) entry.actor = opts.actor;
  if (opts.details !== undefined) entry.details = opts.details;
  return entry;
}
